File: lib/generate_embeddings.py
#!/usr/bin/env python3
"""
Generate embeddings for fashion dataset using CLIP model.
"""

import argparse
import os
import sys
import importlib.util

# Add error handling for imports
try:
    import numpy as np
    import pandas as pd
    import torch
    import clip
    from PIL import Image
    from tqdm import tqdm
    import faiss
except ImportError as e:
    print(f"Error importing required module: {e}")
    print("Please make sure all dependencies are installed:")
    print("pip install numpy pandas torch torchvision Pillow tqdm faiss-cpu")
    print("pip install git+https://github.com/openai/CLIP.git")
    sys.exit(1)

def parse_args():
    parser = argparse.ArgumentParser(description="Generate embeddings for fashion dataset")
    parser.add_argument("--dataset-path", type=str, required=True, help="Path to dataset directory")
    parser.add_argument("--embeddings-path", type=str, required=True, help="Path to save embeddings")
    parser.add_argument("--batch-size", type=int, default=16, help="Batch size for processing")
    parser.add_argument("--env-file", type=str, help="Path to environment variables file")
    
    return parser.parse_args()

def main():
    args = parse_args()
    
    # Load environment variables if provided
    if args.env_file and os.path.exists(args.env_file):
        print(f"Loading environment variables from {args.env_file}")
        spec = importlib.util.spec_from_file_location("env_module", args.env_file)
        env_module = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(env_module)
        
        # Override args with environment variables if they exist
        if hasattr(env_module, 'DATASET_PATH'):
            args.dataset_path = env_module.DATASET_PATH
        if hasattr(env_module, 'EMBEDDINGS_PATH'):
            args.embeddings_path = env_module.EMBEDDINGS_PATH
    
    # Print Python environment info for debugging
    print(f"Python executable: {sys.executable}")
    print(f"Python version: {sys.version}")
    print(f"Python path: {sys.path}")
    
    # Define paths
    DATASET_PATH = args.dataset_path
    EMBEDDINGS_PATH = args.embeddings_path
    IMAGE_FOLDER = os.path.join(DATASET_PATH, 'images')
    METADATA_FILE = os.path.join(DATASET_PATH, 'styles.csv')
    
    # Check if paths exist
    print(f"Checking paths:")
    print(f"Dataset path: {DATASET_PATH}, exists: {os.path.exists(DATASET_PATH)}")
    print(f"Images folder: {IMAGE_FOLDER}, exists: {os.path.exists(IMAGE_FOLDER)}")
    print(f"Metadata file: {METADATA_FILE}, exists: {os.path.exists(METADATA_FILE)}")
    
    assert os.path.exists(IMAGE_FOLDER), f"Image folder not found: {IMAGE_FOLDER}"
    assert os.path.exists(METADATA_FILE), f"Metadata file not found: {METADATA_FILE}"
    
    # Create embeddings directory if it doesn't exist
    os.makedirs(EMBEDDINGS_PATH, exist_ok=True)
    
    # Load dataset
    print("Loading dataset...")
    df = pd.read_csv(METADATA_FILE, on_bad_lines="skip")
    
    # Remove missing values and ensure corresponding images exist
    df = df.dropna(subset=["id", "productDisplayName"])
    df["id"] = df["id"].astype(str)  # Convert ID to string
    
    # Keep only rows where the image exists
    print("Filtering dataset...")
    df = df[df["id"].apply(lambda x: os.path.exists(os.path.join(IMAGE_FOLDER, f"{x}.jpg")))].reset_index(drop=True)
    
    # Create a text description combining metadata
    df["text_description"] = df.apply(
        lambda row: f"{row['productDisplayName']} - {row['masterCategory']}, {row['subCategory']}, {row['articleType']}, {row['baseColour']}, {row['usage']}, {row['gender']}",
        axis=1
    )
    
    # Display dataset info
    print(f"Dataset size after filtering: {len(df)}")
    
    # Load CLIP model
    print("Loading CLIP model...")
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model, preprocess = clip.load("ViT-B/32", device=device)
    
    # Dictionary to store image embeddings
    print("Generating image embeddings...")
    image_embeddings = {}
    
    # Process images in batches
    batch_size = args.batch_size
    for i in tqdm(range(0, len(df), batch_size)):
        batch_df = df.iloc[i:i+batch_size]
        batch_images = []
        batch_ids = []
        
        for _, row in batch_df.iterrows():
            img_id = row["id"]
            img_path = os.path.join(IMAGE_FOLDER, f"{img_id}.jpg")
            
            try:
                image = preprocess(Image.open(img_path).convert("RGB"))
                batch_images.append(image)
                batch_ids.append(img_id)
            except Exception as e:
                print(f"Error processing image {img_id}: {e}")
        
        if batch_images:
            images_tensor = torch.stack(batch_images).to(device)
            
            with torch.no_grad():
                image_features = model.encode_image(images_tensor)
                image_features /= image_features.norm(dim=-1, keepdim=True)
            
            for img_id, feature in zip(batch_ids, image_features):
                image_embeddings[img_id] = feature.cpu().numpy()
    
    # Save image embeddings
    print("Saving image embeddings...")
    np.save(os.path.join(EMBEDDINGS_PATH, "image_embeddings.npy"), image_embeddings)
    
    # Generate text embeddings
    print("Generating text embeddings...")
    text_embeddings = {}
    
    for i in tqdm(range(0, len(df), batch_size)):
        batch_df = df.iloc[i:i+batch_size]
        batch_texts = []
        batch_ids = []
        
        for _, row in batch_df.iterrows():
            img_id = row["id"]
            text = row["text_description"]
            
            batch_texts.append(text)
            batch_ids.append(img_id)
        
        if batch_texts:
            text_tokens = clip.tokenize(batch_texts).to(device)
            
            with torch.no_grad():
                text_features = model.encode_text(text_tokens)
                text_features /= text_features.norm(dim=-1, keepdim=True)
            
            for img_id, feature in zip(batch_ids, text_features):
                text_embeddings[img_id] = feature.cpu().numpy()
    
    # Save text embeddings
    print("Saving text embeddings...")
    np.save(os.path.join(EMBEDDINGS_PATH, "text_embeddings.npy"), text_embeddings)
    
    # Create FAISS index
    print("Creating FAISS index...")
    image_vectors = np.array(list(image_embeddings.values()))
    dimension = image_vectors.shape[1]
    index = faiss.IndexFlatL2(dimension)
    index.add(image_vectors)
    
    # Save FAISS index
    print("Saving FAISS index...")
    faiss.write_index(index, os.path.join(EMBEDDINGS_PATH, "fashion_faiss.index"))
    
    print("Embeddings and index generated successfully.")

if __name__ == "__main__":
    main()

File: scripts/generate-embeddings.js
const { exec } = require("child_process")
const { promisify } = require("util")
const path = require("path")
const fs = require("fs")
const dotenv = require("dotenv")

// Load environment variables from .env.local
const envPath = path.join(process.cwd(), ".env.local")
if (fs.existsSync(envPath)) {
  console.log(`Loading environment variables from ${envPath}`)
  dotenv.config({ path: envPath })
}

const execAsync = promisify(exec)

// Define paths
const DATASET_PATH = process.env.DATASET_PATH || "./data/fashion-dataset"
const EMBEDDINGS_PATH = path.join(DATASET_PATH, "embeddings")
const SCRIPT_PATH = path.join(process.cwd(), "lib", "generate_embeddings.py")

// Get the path to the virtual environment's Python interpreter
const VENV_PATH = path.join(process.cwd(), ".venv")
const PYTHON_EXECUTABLE = path.join(VENV_PATH, "Scripts", "python.exe")

async function generateEmbeddings() {
  try {
    console.log("Checking if dataset exists...")
    console.log(`Dataset path: ${DATASET_PATH}`)

    if (!fs.existsSync(DATASET_PATH)) {
      console.error(`Dataset directory not found at ${DATASET_PATH}`)
      return false
    }

    // Check for images directory and metadata file
    const imagesPath = path.join(DATASET_PATH, "images")
    const metadataPath = path.join(DATASET_PATH, "styles.csv")

    if (!fs.existsSync(imagesPath) || !fs.existsSync(metadataPath)) {
      console.error("Dataset files not found. Please download the dataset manually.")
      return false
    }

    // Create embeddings directory if it doesn't exist
    if (!fs.existsSync(EMBEDDINGS_PATH)) {
      console.log("Creating embeddings directory...")
      fs.mkdirSync(EMBEDDINGS_PATH, { recursive: true })
    }

    // Check if embeddings already exist
    const imageEmbeddingsPath = path.join(EMBEDDINGS_PATH, "image_embeddings.npy")
    const textEmbeddingsPath = path.join(EMBEDDINGS_PATH, "text_embeddings.npy")
    const faissIndexPath = path.join(EMBEDDINGS_PATH, "fashion_faiss.index")

    if (fs.existsSync(imageEmbeddingsPath) && fs.existsSync(textEmbeddingsPath) && fs.existsSync(faissIndexPath)) {
      console.log("Embeddings already exist. Skipping generation.")
      return true
    }

    console.log("Generating embeddings. This may take a while...")
    console.log(`Using Python interpreter: ${PYTHON_EXECUTABLE}`)

    // Create a temporary .env file for the Python script
    const tempEnvPath = path.join(process.cwd(), "temp_env.py")
    fs.writeFileSync(
      tempEnvPath,
      `
DATASET_PATH = "${DATASET_PATH.replace(/\\/g, "\\\\")}"
EMBEDDINGS_PATH = "${EMBEDDINGS_PATH.replace(/\\/g, "\\\\")}"
    `,
    )

    // Execute Python script with environment variables
    const command = `"${PYTHON_EXECUTABLE}" "${SCRIPT_PATH}" --dataset-path "${DATASET_PATH}" --embeddings-path "${EMBEDDINGS_PATH}" --env-file "${tempEnvPath}"`
    console.log(`Running command: ${command}`)

    const { stdout, stderr } = await execAsync(command)

    // Clean up temp file
    if (fs.existsSync(tempEnvPath)) {
      fs.unlinkSync(tempEnvPath)
    }

    if (stderr) {
      console.error(`Python Error: ${stderr}`)
    }

    console.log(stdout)
    console.log("Embeddings generated successfully.")
    return true
  } catch (error) {
    console.error("Error generating embeddings:", error)
    return false
  }
}

// Run the function if this file is executed directly
if (require.main === module) {
  generateEmbeddings().then((success) => {
    if (success) {
      console.log("Embeddings generation completed successfully.")
    } else {
      console.error("Embeddings generation failed.")
    }
  })
}

module.exports = generateEmbeddings

